examples/librilight/matching.py

```diff
@@ -0,0 +1,906 @@
+            # alignments in a supervision might be empty
+            if aligns["text"]:
+                transcript = Transcript.from_dict(
+                    name=sup.id,
+                    d=aligns,
+                    use_utf8=params.use_utf8,
+                    is_bpe=params.is_bpe,
+                )
+                query_len += transcript.binary_text.size
+                transcripts.append(transcript)
+                transcripts_cut_index.append((i, j))
+        book_paths.add(cut.text_path)
+
+    # Construct references (the books)
+    for i, book_path in enumerate(book_paths):
+        with open(params.book_dir / book_path, "r") as f:
+            book_text = f.read()
+            book = TextSource.from_str(
+                name=book_path,
+                s=book_text,
+                use_utf8=params.use_utf8,
+            )
+            books.append(book)
+
+    logging.info("Loading data done.")
+
+    sourced_transcript_lists = texts_to_sourced_texts(
+        transcripts, uppercase=params.use_uppercase
+    )
+    sourced_transcripts = append_texts(sourced_transcript_lists)
+
+    sourced_book_list = texts_to_sourced_texts(books, uppercase=params.use_uppercase)
+    sourced_books = append_texts(sourced_book_list)
+
+    def _is_not_punc(c: np.int32) -> bool:
+        return c not in PUNCTUATION
+
+    # Removing the punctuation
+    sourced_books = filter_texts(sourced_books, fn=_is_not_punc)
+
+    sourced_text = append_texts([sourced_transcripts, sourced_books])
+
+    assert query_len == sourced_text.doc_splits[len(transcripts)], (
+        query_len,
+        sourced_text.doc_splits[len(transcripts)],
+    )
+
+    logging.info(f"Creating suffix array.")
+    suffix_array = create_suffix_array(sourced_text.binary_text)
+
+    logging.info(f"Finding close matches.")
+    close_matches = find_close_matches(
+        suffix_array, query_len, num_close_matches=params.num_close_matches
+    )
+
+    logging.info(f"Finding candidates.")
+    candidate_matches = find_candidate_matches(
+        close_matches=close_matches,
+        text=sourced_text,
+        num_candidates=params.num_candidates,
+        length_ratio=params.match_length_ratio,
+    )
+
+    assert len(transcripts) == len(candidate_matches), (
+        len(transcripts),
+        len(candidate_matches),
+    )
+
+    # Prepare arguments for levenshtein which would run in parallel
+    # Read https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.ThreadPool
+    # for more details
+    arguments = []
+    for i in range(len(transcripts)):
+        arguments.append(
+            (i, sourced_text, transcripts_cut_index[i], candidate_matches[i])
+        )
+
+    # Note: Only the levenshtein part runs in the ThreadPool.
+    # other parts are implemented in python, we can not speed them up due to
+    # the GIL
+    num_processes = min(len(arguments), os.cpu_count())
+    with Pool(num_processes) as pool:
```

Comment:
Instead of re-creating a pool of processes for each batch, I suggest that you create only a pool of processes once.

Process creation takes time and you don't need to recreate them for each batch.